### PR TITLE
fix(query): fix inner type of nested type can not be set as not null

### LINKED ISSUE #13235: bug: inner type of nested type can not be set as not null
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

v1.2.154-nightly

### What's Wrong?

inner type of nested type can not be set as not null

### How to Reproduce?

```sql
mysql> create table test(a int, b array(int not null));
ERROR 1105 (HY000): SyntaxException. Code: 1005, Text = error:
  --> SQL:1:38
  |
1 | create table test(a int, b array(int not null))
  | ------                   - -----     ^^^ expected `NULL`, `UNSIGNED`, `)`, or `(`
  | |                        | |
  | |                        | while parsing type name
  | |                        while parsing `<column name> <type> [DEFAULT <expr>] [AS (<expr>) VIRTUAL] [AS (<expr>) STORED] [COMMENT '<comment>']`
  | while parsing `CREATE TABLE [IF NOT EXISTS] [<database>.]<ta
```

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by b41sh**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

introcude `TypeName::NotNull(Box<TypeName>)` for inner not null types.
fix inner type of nested type can not be set as not null.

- Fixes #13235

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [x] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14379)

### CODE REVIEW

File: tests/sqllogictests/suites/mode/standalone/explain_native/project_set.test [hunk 1]
```diff
@@ -26,7 +26,7 @@ AggregateFinal
     └── ProjectSet
         ├── output columns: [unnest (#2)]
         ├── estimated rows: 1.00
-        ├── set returning functions: unnest(CAST(fold_count.id (#0) AS Array(Int32 NULL) NULL))
+        ├── set returning functions: unnest(fold_count.id (#0))
```
Comment (andylokandy) [inline tests/sqllogictests/suites/mode/standalone/explain_native/project_set.test:29]: is this expected?

File: src/query/ast/src/parser/expr.rs [hunk 2]
```diff
@@ -1603,14 +1603,23 @@ pub fn type_name(i: Input) -> IResult<TypeName> {
             | #ty_string
             | #ty_variant
             | #ty_nullable
-            ) ~ NULL? : "type name" },
+            ) ~ NOT? ~ NULL? : "type name" },
```
Comment (andylokandy) [inline src/query/ast/src/parser/expr.rs:1606]:
```suggestion
    let nullable = alt((
        value(true, rule! { NULL }),
        value(false, rule! { NOT ~ ^NULL }),
    ));
    map(
        alt((
            rule! {
            ( #ty_boolean
            | #ty_uint8
            | #ty_uint16
            | #ty_uint32
            | #ty_uint64
            | #ty_int8
            | #ty_int16
            | #ty_int32
            | #ty_int64
            | #ty_float32
            | #ty_float64
            | #ty_decimal
            | #ty_array
            | #ty_map
            | #ty_bitmap
            | #ty_tuple : "TUPLE(<type>, ...)"
            | #ty_named_tuple : "TUPLE(<name> <type>, ...)"
            ) ~ nullable? : "type name"
            },
            rule! {
            ( #ty_date
            | #ty_datetime
            | #ty_binary
            | #ty_string
            | #ty_variant
            | #ty_nullable
            ) ~ nullable? : "type name" },
        )),
        |(ty, opt_nullable)| {
            match opt_nullable {
                Some(true) => ty.wrap_nullable(),
                Some(false) => ty.wrap_not_null(),
                None => ty,
            }
        },
    )(i)
```

File: tests/sqllogictests/suites/mode/standalone/explain_native/project_set.test (hunk 1 above)
Comment (b41sh) [reply tests/sqllogictests/suites/mode/standalone/explain_native/project_set.test:29]: yes, the data type is `Array(Int32 NULL) NULL`, so cast is not needed

File: src/query/ast/src/parser/expr.rs [hunk 3]
```diff
@@ -1603,12 +1610,11 @@ pub fn type_name(i: Input) -> IResult<TypeName> {
...
+            ) ~ #nullable? : "type name" },
         )),
-        |(ty, opt_not, opt_null)| match (opt_not.is_some(), opt_null.is_some()) {
-            (false, true) => Ok(ty.wrap_nullable()),
-            (false, false) => Ok(ty),
-            (true, true) => {
+        |(ty, opt_nullable)| match opt_nullable {
+            Some(true) => Ok(ty.wrap_nullable()),
+            Some(false) => {
                 if matches!(ty, TypeName::Nullable(_)) {
```
Comment (andylokandy) [inline src/query/ast/src/parser/expr.rs:1618]: this is impossible because parser of `ty` contains no nullable

Comment (b41sh) [reply src/query/ast/src/parser/expr.rs:1618]: We also support `nullable(type)` syntax
for example
```sql
CREATE TABLE t(c1 NULLABLE(int));
```
in this case `c1 NULLABLE(int) NOT NULL` is invalid

**Review by andylokandy — APPROVED** [review]

**Pull request merged**